### PR TITLE
passing metadata to reduce handler

### PR DESCRIPTION
**Pull request by yhl25**
Fixes https://github.com/numaproj/numaflow/issues/334

Signed-off-by: Yashash H L <yashash_hl@intuit.com>

### CODE REVIEW

File: pkg/function/service.go [hunk 1]
```diff
@@ -81,26 +117,47 @@ func (fs *Service) MapFn(ctx context.Context, d *functionpb.Datum) (*functionpb.
...
+	if err != nil {
+		return err
+	}
+
+	et, err = getValueForKey(grpcMD, WinEndTime)
+	if err != nil {
+		return err
 	}
 
+	startTime, _ = strconv.Atoi(st)
```
Comment (ashwinidulams) [inline pkg/function/service.go:152]: should use strconv.ParseInt functions here for int64 values.

File: pkg/function/service.go [hunk 2]
```diff
@@ -81,26 +117,47 @@ func (fs *Service) MapFn(ctx context.Context, d *functionpb.Datum) (*functionpb.
...
+	endTime, _ = strconv.Atoi(et)
+
+	// create interval window interface using the start and end time
+	iw := NewIntervalWindow(time.UnixMilli(int64(startTime)), time.UnixMilli(int64(endTime)))
+
+	// create metadata using interval window interface
+	md = NewMetadata(iw)
```
Comment (ashwinidulams) [inline pkg/function/service.go:159]: do we need Metadata? why dont we make it explicit and pass in IntervalWindow as an argument?
we dont expect any additional metadata here.

File: pkg/function/service.go [hunk 3]
```diff
@@ -142,3 +199,21 @@ func (fs *Service) ReduceFn(stream functionpb.UserDefinedFunction_ReduceFnServer
...
 	})
 }
+
+func getValueForKey(md grpcmd.MD, key string) (string, error) {
+	var value string
+
+	keyValue := md.Get(key)
+
+	if len(keyValue) > 1 {
+		return value, fmt.Errorf("expected extactly one value for key %s in metadata but got %d values", key, len(keyValue))
```
Comment (ashwinidulams) [inline pkg/function/service.go:209]: log values as well.

File: pkg/function/service.go (hunk 1 above)
Comment (yhl25) [reply pkg/function/service.go]: yes, done

File: pkg/function/service.go (hunk 3 above)
Comment (yhl25) [reply pkg/function/service.go]: done

**Review by vigith — APPROVED** [review]
LTGM

File: pkg/function/service.go (hunk 2 above)
Comment (chromevoid) [reply pkg/function/service.go:159]: I prefer Metadata. If we are to have more metadata fields in the future, we can directly add in the `md` and the user doesn't have to update their reduce handler signature. e.g. https://github.com/numaproj/numaflow-go/blob/main/pkg/function/examples/sum/main.go#L12

**Review by chromevoid — APPROVED** [review]

**Pull request merged**